### PR TITLE
Sentry Unified Node.js SDK upgrade

### LINKED ISSUE #12: Unified Node.js SDK support planed?
https://docs.sentry.io/clients/node/

Deprecated Client raven-node

This SDK has been superseded by a new unified one. The documentation here is preserved for customers using the old client. For new projects have a look at the new client documentation: [Unified Node.js SDK.](https://docs.sentry.io/quickstart?platform=node)

### PR DESCRIPTION
**Pull request by aheissenberger**
Replace depricated Sentry Raven SDK with current Sentry Unified Node.js SDK
fixes https://github.com/arabold/serverless-sentry-lib/issues/12

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -66,63 +52,61 @@ class _ {
...
+	);
 
-	// Register this instance globally for backward compatibility
-	// with serverless-sentry-plugin 0.2.x/0.3.x
-	global.sls_raven = Raven;
-	ravenInstalled = true;
+	sentryInstalled = true;
 
-	console.log("Raven installed.");
+	console.log("Sentry installed.");
```
Comment (Vadorequest) [inline src/index.js:147]: I'd be more explicit with logs and namespace them. It's not easy for a developer to understand athat this comes from the plugin and not sentry SDK itself.

File: src/index.js [hunk 2]
```diff
@@ -137,41 +121,69 @@ function installTimers(pluginConfig, lambdaContext) {
...
+		if (sentryInstalled) {
+			Sentry.captureMessage("Function Execution Time Warning", {
+				level: "warning",
+				extra: {
+					TimeRemainingInMsec: lambdaContext.getRemainingTimeInMillis()
+				}
+			});
+			const client = Sentry.getCurrentHub().getClient();
+			if (client) {
+				client.flush(1000).then(function() {
```
Comment (Vadorequest) [inline src/index.js]: This seems new and could use some comments explaining what it actually does

File: README.md [hunk 3]
```diff
@@ -8,14 +8,15 @@
...
 support will require dedicated libraries. Pull requests are welcome!
 
 ### What is Raven and Sentry?
-It's a bit confusing, but _Raven_ is the official name of the error reporting SDK
-that will forward errors, exceptions and messages to the _Sentry_ server. For
-more details of what Raven and Sentry actually is, refer to the official Sentry documentation: https://docs.sentry.io/.
+It's a bit confusing, but _Raven_ *was* the old official name of the error reporting SDK
+that will forward errors, exceptions and messages to the _Sentry_ server. The new name is
+_Unified Node.js SDK_. 
+For more details of what Raven and Sentry actually is, refer to the official Sentry documentation: https://docs.sentry.io/.
```
Comment (Vadorequest) [inline README.md:21]: I'd maybe add something stating the old Raven SDK is still available using v1.1.2? Maybe some people would still look for the old implementation.

File: src/index.js (hunk 2 above)
Comment (aheissenberger) [reply src/index.js]: captureMessage & captureException are no promises nor do they provide a callback when finished with the transfer to sentry.io. This tells the Hub client to flush all open events and wait max 1000ms or return when finished.

Comment (aheissenberger) [reply src/index.js]: found a newer/shorter version of flush :-)

File: README.md (hunk 3 above)
Comment (aheissenberger) [reply README.md:21]: updated the Readme to point to old Version

File: src/index.js [hunk 4]
```diff
@@ -366,82 +450,70 @@ class RavenLambdaWrapper {
...
-					captureUnhandled(err);
-				}
-			});
+				// Returning non-Promise values would be meaningless for lambda.
+				// But inherit the behavior of the original handler.
+				return promise;
+			}
+			catch (err) {
+				// Catch and log synchronous exceptions thrown by the handler
+				captureUnhandled(err);
```
Comment (ahanoff) [inline src/index.js:513]: @aheissenberger  I'm testing it locally, and somehow my errors are not captured here.

Comment (ahanoff) [reply src/index.js:513]: Shall captureUnhandled be in line 504 too?

Comment (aheissenberger) [reply src/index.js:513]: @ahanof 

> Shall captureUnhandled be in line 504 too?

no - the capture happens in the wrapper of the fail handler line 322 and 452.
Serverless offline or AWS Lambda will do a callback on "fail" when rejected.
Have a look at the tests to see what happens and to find out what's the problem here.

Do you use promises or callbacks in your lambda function?

Comment (ahanoff) [reply src/index.js:513]: I'm using promises. I'm testing locally with yalc. Problem is that debugger is not stopping with breakpoint at wrapCallback (line 262) 🤔

**Comment by ahanoff** [comment]
@aheissenberger ok it definitely doesn't work. I do simple revert to Raven, and it immediately sends error to sentry. If I switch to your branch and sentryConfig, events are not showing up in sentry.

Let me share my code with both: old raven and new sentry/node

**Comment by ahanoff** [comment]
https://github.com/ahanoff/sentry-test so here my simple aws lambda test, so your code does work with callbacks, but somehow it doesn't work with async/await. Update: you need your own dsn in serverless.yml

**Comment by tobiasvoelke** [comment]
Any update here? Would highly appreciate this PR.
Let me know if I can support

**Comment by aheissenberger** [comment]
> Any update here? Would highly appreciate this PR.
> Let me know if I can support

sorry @tobiasvoelke but I am currently blocked by other tasks and will not finish this before Ende of July

**Comment by Vadorequest** [comment]
Any update on this? Any help needed?

**Comment by ahanoff** [comment]
@Vadorequest  can you test and verify if it works with async/await lambda? Maybe I'm doing something wrong, but it doesn't work for me

**Comment by Vadorequest** [comment]
I'm in holidays with little time available, is there a reproduction I could just fork and test to save time?

**Comment by ahanoff** [comment]
@aheissenberger  sorry, my bad, it actually works with async/await. In my exampla I was using https://github.com/middyjs and I think I messed up with error handling there.

When this pull request can be merged?

**Comment by azai91** [comment]
Any idea on when this will be merged?

**Comment by polosatyi** [comment]
Any updates?

**Comment by polosatyi** [comment]
Any updates?

**Comment by arabold** [comment]
Sorry for the total lack of communication from my side. I'm currently working on a major rewrite v2.0.0 based on this PR and some more changes. I'll push an update as soon as I can.

**Comment by arabold** [comment]
Closing this. v2.0.0 has been pushed and is being tested. A new `npm` update coming soon as well.

**Pull request closed without merging**